### PR TITLE
numpy vs pil simple memory usage

### PR DESCRIPTION
**Pull request by qiminchen**
This is a simple memory usage comparison of `Numpy` and `PIL` mentioned at https://github.com/beijbom/pyspacer/pull/17#discussion_r420909318

### CODE REVIEW

**Comment by qiminchen** [comment]
@beijbom hey Oscar, as you mentioned at https://github.com/beijbom/pyspacer/pull/17#discussion_r420909318, one way to save more memory is to crop the image first using `Image.crop()` from `PIL` and convert the patches to `Numpy` after. So I did a very simple experiment to trace memory usage. Method for monitoring the memory usage adopted from [Option 2: tracemalloc](https://medium.com/survata-engineering-blog/monitoring-memory-usage-of-a-running-python-program-49f027e3d1ba) and refer to [tracemalloc](https://docs.python.org/3/library/tracemalloc.html#tracemalloc.get_traced_memory) for more details.

Basically, `crop_patches_pil()` in `extract_features_utils.py` return a list of cropped PIL Image, `if...else...` in `extract_features.py` just make the comparison convenient. `TestMemoryUsage` in `test_extract_features.py` simply test the memory usage of extraction block. These auxiliary functions will be removed if we decide to merge this PR.

`tracemalloc` basically tracks the individual memory blocks allocated by the Python interpreter when `__call__(self, im, rowcols)` in `EfficientNetExtractor()` is completed so I assume this is equivalent to the memory usage measurement.

Result on `08bfc10v7t.png` with 5 annotations. Looks like cropping the image first using `Image.crop()` from `PIL` and convert the patches to `Numpy` after would save more memory than converting to numpy array at the beginning.
```
-> Initializing EfficientNetExtractor
Numpy: Current memory usage is 0.464665MB; Peak was 4.160963MB
-> Initializing EfficientNetExtractor
PIL: Current memory usage is 0.46915MB; Peak was 2.939177MB
```

File: spacer/extract_features_utils.py [hunk 1]
```diff
@@ -60,3 +60,20 @@ def crop_simple(im: np.ndarray,
...
+                     rowcols: List[Tuple[int, int]],
+                     crop_size: int):
+    pad = crop_size
+    image = Image.fromarray(
+        np.pad(im.convert('RGB'),
+               ((pad, pad), (pad, pad), (0, 0)), mode='reflect'),
+        'RGB'
+    )
+
+    return [image.crop((pad + col - crop_size / 2,
```
Comment (beijbom) [inline spacer/extract_features_utils.py:75]: Image.crop returns Image objects, so you'd need to convert to arrays here to match with the previous method. https://pillow.readthedocs.io/en/3.1.x/reference/Image.html#PIL.Image.Image.crop

**Comment by beijbom** [comment]
hey @qiminchen . Thanks for looking into this. It's interesting it saves so much memory since you are still converting it to an array in order to do the np.pad call. How much does this save on a 100 mega pixel image?

I'm inclined to stay with the old code since we know that it works (don't fix what ain't broken...). Also, I have had some bad experience with the PIL.convert('RGB') step -- just seem to cause trouble for some of the more unusual image formats. 

But let's keep an eye on this fix if we run into memory issues down the road.

File: spacer/extract_features_utils.py (hunk 1 above)
Comment (qiminchen) [reply spacer/extract_features_utils.py:75]: if we convert to arrays here, wouldn't it be the same as the original `crop_patches`?

In Caffe, we have `Transformer` that can process `Image` object https://github.com/beijbom/pyspacer/blob/b106de785e21e14cfbac03878ff85f1b48b80847/spacer/caffe_utils.py#L26-L34
In Pytorch, we have `transformation` that can process `Image` object as well
https://github.com/beijbom/pyspacer/blob/b106de785e21e14cfbac03878ff85f1b48b80847/spacer/torch_utils.py#L15-L24

**Comment by qiminchen** [comment]
@beijbom 
> How much does this save on a 100 mega pixel image?

I will take a look at this after.
> I'm inclined to stay with the old code since we know that it works (don't fix what ain't broken...). Also, I have had some bad experience with the PIL.convert('RGB') step -- just seem to cause trouble for some of the more unusual image formats.

I agree.

**Comment by beijbom** [comment]
@qiminchen : I'm closing this PR for now. We can revisit later.

**Pull request closed without merging**